### DISCUSSION TITLE
`None` comparison is not possible in new versions

### DISCUSSION THREAD

**Discussion by bloussou** [microsoft/pyright#3688] [Q&A] [answered]
Hello,
In `1.1.224` it was possible to have the following code but it is not possible with `1.1.258`. Is this an expected behavior ?
```python
from typing import Any

class B:
    pass

class A:
    def __eq__(self, other: Any) -> B:  # type: ignore[override]
        print("b")
        return B()
# It was not failing before because I think it was inferring the __eq__ was overridden
if A() == None: # Condition will always evaluate to False since the types "A" and "None" have no overlap
    ...
```

**Comment by erictraut** [comment, marked as answer]
The current behavior is "as designed" even if it's not entirely expected.

I see that you have enabled the strict-mode check `reportUnnecessaryComparison`. This check needs to make some assumptions about the way comparisons work. These assumptions work in most use cases, but there are edge cases where the heuristic will fail. 

You're overriding `__eq__` in a very unusual way in this code. In fact, you're overriding the `__eq__` method for `object` in an incompatible manner and then ignoring the error using a `# type: ignore` comment.

I've fixed many bugs and made other changes since version `1.1.224`, so I can't say specifically what caused this to work previously. If you can find the specific version where this behavior changed, I'm willing to investigate further.

**Reply by bloussou** [reply]
Hello,
Thanks for your answer.
This behavior changed in version `1.1.129`

**Reply by erictraut** [reply]
Thanks. Yes, I introduced a bug fix in 1.1.129 that would explain why you didn't see this behavior prior to that.

The check is working as expected. The assumption is that a comparison of some class instance with `None` will always produce a `False` result. Of course, with Python you can always override `__eq__` and `__ne__` in weird and unusual ways, but checks like this one need to make assumptions that they will be used in a standard manner. If you find that the `reportUnnecessaryComparison` check doesn't meet your needs, you can disable it.

**Reply by bloussou** [reply]
indeed, we should think of something like pandas `isnull()` probably. Thanks a lot for your explanations.
